src/hooks/useDialog.js

```diff
@@ -0,0 +1,84 @@
+import {useCallback, useEffect} from 'react'
+
+function useDialog({modalRef, isOpen, onDismiss, initialFocusRef, closeButtonRef, returnFocusRef} = {}) {
+  const onClickOutside = useCallback(
+    (e) => {
+      if (modalRef && modalRef.current && !modalRef.current.contains(e.target)) {
+        onDismiss()
+        if (returnFocusRef && returnFocusRef.current) {
+          returnFocusRef.current.focus()
+        }
+      }
+    },
+    [onDismiss, modalRef, returnFocusRef]
+  )
+
+  useEffect(() => {
+    if (isOpen) {
+      document.addEventListener('click', onClickOutside)
+      return () => {
+        document.removeEventListener('click', onClickOutside)
+      }
+    }
+  }, [isOpen, onClickOutside])
+
+  useEffect(() => {
+    if (isOpen) {
+      if (initialFocusRef && initialFocusRef.current) {
+        initialFocusRef.current.focus()
+      } else if (closeButtonRef && closeButtonRef.current) {
+        closeButtonRef.current.focus()
+      }
+    }
+  }, [isOpen, initialFocusRef, closeButtonRef, returnFocusRef])
+
+  const getFocusableItem = (e, movement) => {
+    if (modalRef && modalRef.current) {
+      const items = Array.from(modalRef.current.querySelectorAll('*')).filter(focusable)
+      if (items.length === 0) return
+      e.preventDefault()
+
+      const focusedElement = document.activeElement
+      const index = items.indexOf(focusedElement)
+      const offsetIndex = index + movement
+      const fallbackIndex = movement === 1 ? 0 : items.length - 1
+      return items[offsetIndex] || items[fallbackIndex]
+    }
+  }
+
+  function visible(el) {
+    return !el.hidden && (!el.type || el.type !== 'hidden') && (el.offsetWidth > 0 || el.offsetHeight > 0)
+  }
```

Comment:
Can we pull this function definition out of the hook? It doesn't look like it depends on anything within the hook's function scope.

Not sure if we have a utilities module or not, but just a non-exported function in `useDialog.js` is probably fine as well.

Reply:
Sounds good! 👍